pkg/platform/runtime/buildscript/json.go

```diff
@@ -40,7 +41,17 @@ func (v *Value) MarshalJSON() ([]byte, error) {
 	case v.FuncCall != nil:
 		return json.Marshal(v.FuncCall)
 	case v.List != nil:
-		return json.Marshal(v.List)
+		// Buildexpression list order does not matter, so sorting is necessary for
+		// comparisons. Go's JSON marshaling is deterministic, so utilize that.
+		// This should not be necessary when DX-1939 is implemented.
+		list := make([]*Value, len(*v.List))
+		copy(list, *v.List)
+		sort.SliceStable(list, func(i, j int) bool {
+			b1, err1 := json.Marshal(list[i])
+			b2, err2 := json.Marshal(list[j])
+			return err1 == nil && err2 == nil && string(b1) < string(b2)
+		})
```

Comment:
This feels too hacky. We are effectively throwing our hands up and hoping for the best if an error occurs when marshalling. We're also marshalling list values redundantly, which I imagine could add up significantly for build expressions.

Is this still warranted considering you have `getComparableJson`? Can we handle this use-case differently?

Reply:
> This feels too hacky. We are effectively throwing our hands up and hoping for the best if an error occurs when marshalling. We're also marshalling list values redundantly, which I imagine could add up significantly for build expressions.

No, this is a stable sort, so if there are errors marshaling (and there really shouldn't be), the order will remain the same. If that were to happen (and I cannot see how it would), then the worst that happens is that buildscript equality fails and we do a merge. That seems safe.

The redundant marshaling is happening in Go, so it's going to be fast. Moreover, this only happens when checking build script equality. I think the network call to fetch the remote build expression takes longer than the actual equality check.

> Is this still warranted considering you have `getComparableJson`? Can we handle this use-case differently?

`getComparableJson()` invokes this function because I need to handle different list order (https://github.com/ActiveState/cli/pull/2582/commits/d307bf98f29fc10557c8843f9d9a3a4acc214753#diff-1ebd8e286e3a9885c33bb17b9fdb14f4d2624f1d9dda84792bb018dd3352866eR171-R178). This will go away once we have hash checking in DX-1939.

Since we are currently comparing JSON directly at this time, we need everything to be ordered properly. Maps are fine, but lists are not. The only other way around it is to have a BuildScript struct method written in Go that iterates over all members, checking for equality with logic to handle out of order things. That's more throwaway code than this marshaling & string comparison method.